File: lib/circle.rb
require_relative 'point'

class Circle
  
  attr_accessor :center, :radius

  def initialize center, radius
    @center = center
    @radius = radius.to_f
  end

  def == other
    @center == other.center and @radius == other.radius
  end

  def circumscribes? point
    squared_dist = (@center.x - point.x) ** 2.0 + (center.y - point.y) ** 2.0
    squared_dist <= @radius ** 2.0
  end

end

File: lib/delaunay/circle.c
#include "circle.h"

Circle*
circle_new(Point *center, double radius)
{
	Circle *c;
	c = (Circle*)malloc(sizeof(Circle));
	c->center = (Point*)malloc(sizeof(Point));
	memcpy(
		c->center,
		center,
		sizeof(Point)
	);
	c->radius = radius;
	return c;
}

int
circle_equals(Circle *c1, Circle *c2)
{
	if (point_equals(c1->center, c2->center) && c1->radius == c2->radius) return 1;
	return 0;
}

int
circle_circumscribes(Circle *c, Point *p) 
{
	double squared_dist = pow(c->center->x - p->x, 2) +
	                      pow(c->center->y - p->y, 2);
	if (squared_dist <= pow(c->radius, 2)) return 1;
	return 0;
}

void
circle_free(Circle *c)
{
	point_free(c->center);
	free(c);
}
